contrib/golang/filters/http/source/golang_filter.cc

```diff
@@ -0,0 +1,821 @@
+  ENVOY_LOG(debug, "golang filter onLocalReply, state: {}, phase: {}, code: {}", state.stateStr(),
+            state.phaseStr(), int(data.code_));
+
+  return Http::LocalErrorStatus::Continue;
+}
+
+Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers, bool end_stream) {
+  ProcessorState& state = decoding_state_;
+
+  ENVOY_LOG(debug, "golang filter decodeHeaders, state: {}, phase: {}, end_stream: {}",
+            state.stateStr(), state.phaseStr(), end_stream);
+
+  state.setEndStream(end_stream);
+
+  bool done = doHeaders(state, headers, end_stream);
+
+  return done ? Http::FilterHeadersStatus::Continue : Http::FilterHeadersStatus::StopIteration;
+}
+
+Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_stream) {
+  ProcessorState& state = decoding_state_;
+  ENVOY_LOG(debug,
+            "golang filter decodeData, state: {}, phase: {}, data length: {}, end_stream: {}",
+            state.stateStr(), state.phaseStr(), data.length(), end_stream);
+
+  state.setEndStream(end_stream);
+
+  bool done = doData(state, data, end_stream);
+
+  if (done) {
+    state.doDataList.moveOut(data);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  return Http::FilterDataStatus::StopIterationNoBuffer;
+}
+
+Http::FilterTrailersStatus Filter::decodeTrailers(Http::RequestTrailerMap& trailers) {
+  ProcessorState& state = decoding_state_;
+  ENVOY_LOG(debug, "golang filter decodeTrailers, state: {}, phase: {}", state.stateStr(),
+            state.phaseStr());
+
+  state.setSeenTrailers();
+
+  bool done = doTrailer(state, trailers);
+
+  return done ? Http::FilterTrailersStatus::Continue : Http::FilterTrailersStatus::StopIteration;
+}
+
+Http::FilterHeadersStatus Filter::encodeHeaders(Http::ResponseHeaderMap& headers, bool end_stream) {
+  ProcessorState& state = getProcessorState();
+  ENVOY_LOG(debug, "golang filter encodeHeaders, state: {}, phase: {}, end_stream: {}",
+            state.stateStr(), state.phaseStr(), end_stream);
+
+  encoding_state_.setEndStream(end_stream);
+
+  // NP: may enter encodeHeaders in any phase & any state_,
+  // since other filters or filtermanager could call encodeHeaders or sendLocalReply in any time.
+  // eg. filtermanager may invoke sendLocalReply, when scheme is invalid,
+  // with "Sending local reply with details // http1.invalid_scheme" details.
+  if (state.state() != FilterState::Done) {
+    ENVOY_LOG(warn,
+              "golang filter enter encodeHeaders early, maybe sendLocalReply or encodeHeaders "
+              "happened, current state: {}, phase: {}",
+              state.stateStr(), state.phaseStr());
+
+    ENVOY_LOG(debug, "golang filter drain data buffer since enter encodeHeaders early");
+    // NP: is safe to overwrite it since go code won't read it directly
+    // need drain buffer to enable read when it's high watermark
+    state.drainBufferData();
+
+    // get the state before changing it.
+    bool inGo = state.isProcessingInGo();
```

Comment:
nit: snake_cased here and elsewhere.

```suggestion
    bool in_go = state.isProcessingInGo();
```

Reply:
Done, Thanks.